### ISSUE TITLE
Corrupt collection

### ISSUE THREAD

**Issue by timrae** [ankidroid/Anki-Android#3576]
We occasionally get reports from users that their collections have been corrupted. Unfortunately we have been trying to find the root cause of this for years to no avail. Any information on this will be appreciated

**Comment by okejokej** [comment]
Had this problem a lot, and in my case moving the collection to internal memory (as Ankidroid by default created it on SD card) has resolved it (so far), even though otherwise I have no problems with using the SD card. Though that's probably the first thing people check if they experience the issue.

MT6575M based device with Android 4.1.1

**Comment by bjornmartensson** [comment]
I seem to have a reproducible corruption problem. My repro steps:
1. Modify my collection on my PC in a way that requires a full upload.
2. Sync PC -> AnkiWeb.
3. Sync AnkiWeb -> AnkiDroid

Result:
AnkiDroid has a problem and sends a report somewhere. The collection is somehow corrupt and even though there should be questions due, they don't show up.

This happened two days (2017-02-12) and yesterday (2017-02-13) and I seem to be able to reproduce this any day now. Before Feb 12, this never happened to me.

This is bad for me - I love Anki / AnkiDroid and use it on a daily basis. I would be interested in analyzing the crash report myself if nobody else has time for this. Would it be possible for me to get a login for Acralyzer? I'm guessing this is where my crash reports have gone?

Thanks in advance,
Bjorn

**Comment by timrae** [comment]
How did you handle the corruption issue after each reproduction exactly? To start with please provide your "debug info" and I'll post the stack trace

**Comment by bjornmartensson** [comment]
After each reproduction, I restored to an earlier version in ankidroid, and then synced again. I thought that I fixed the problem by 1) restoring in ankidroid, 2) syncing ankidroid -> ankiweb, but somehow after adding more questions on desktop the collection is interpreted as corrupted in ankidroid again after syncing.

My debug info:
AnkiDroid Version = 2.8.1
Android Version = 5.0.1
ACRA UUID = e224840c-cd94-4e86-a917-b0a2bc92c328

**Comment by timrae** [comment]
Is it possible that the backup which you're attempting to restore from is
also corrupt? Did you trying doing the restoration procedure in the Anki
desktop manual?

**Comment by timrae** [comment]
Did you specifically get a message saying it was corrupt?

The stacktrace does not indicate any corruption:

```bash
java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
at com.ichi2.anki.DeckPicker.showSyncLogMessage(DeckPicker.java:1097)
at com.ichi2.anki.DeckPicker.access$1000(DeckPicker.java:110)
at com.ichi2.anki.DeckPicker$18.onPostExecute(DeckPicker.java:1555)
at com.ichi2.async.Connection.onPostExecute(Connection.java:143)
at com.ichi2.async.Connection.onPostExecute(Connection.java:50)
```

**Comment by timrae** [comment]
Would you be able to open a new issue?

**Comment by bjornmartensson** [comment]
I opened a new issue: #4581.

It's possible that the backup which I attempted to restore from is also corrupt. I'm not sure how to verify this. I tried "optimize database" or something like that on the desktop version and it completed without any complaints.

I don't think I got any message saying that the collection was corrupt, but I have over 150 questions scheduled each day, but after this incident, I had zero questions scheduled for the day except for new entries. Seems to me that the collection got corrupted.
